src/module/custom.rs

```diff
@@ -0,0 +1,279 @@
+        self.impl_into_any()
+    }
+    fn as_any(&self) -> &Any {
+        self.impl_as_any()
+    }
+    fn as_any_mut(&mut self) -> &mut Any {
+        self.impl_as_any_mut()
+    }
+}
+
+/// A raw, unparsed custom section.
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct RawCustomSection {
+    /// This custom section's name.
+    pub name: String,
+
+    /// This custom section's raw data.
+    pub data: Vec<u8>,
+}
+
+impl CustomSection for RawCustomSection {
+    fn name(&self) -> &str {
+        &self.name
+    }
+
+    fn data(&self) -> Cow<[u8]> {
+        self.data.as_slice().into()
+    }
+}
+
+/// The id of some `CustomSection` instance in a `ModuleCustomSections`.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub struct CustomSectionId(Id<Option<Box<dyn CustomSectionAny>>>);
+
+/// The id of a `CustomSection` instance with a statically-known type in a
+/// `ModuleCustomSections`.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub struct TypedCustomSectionId<T: CustomSection> {
+    id: CustomSectionId,
+    _phantom: PhantomData<T>,
+}
+
+impl<T> From<TypedCustomSectionId<T>> for CustomSectionId
+where
+    T: CustomSection,
+{
+    #[inline]
+    fn from(a: TypedCustomSectionId<T>) -> Self {
+        a.id
+    }
+}
+
+impl Tombstone for Option<Box<dyn CustomSectionAny>> {
+    fn on_delete(&mut self) {
+        *self = None;
+    }
+}
+
+/// A collection of custom sections inside a Wasm module.
+///
+/// To add parse and emit your own custom section:
+///
+/// * Define a `MyCustomSection` type to represent your custom section.
+///
+/// * Implement the `walrus::CustomSection` trait for your `MyCustomSection`
+///   type.
+///
+/// * When working with a `walrus::Module` named `my_module`, use
+///   `my_module.customs.take_raw("my_custom_section_name")` to take ownership
+///   of the raw custom section if it is present in the Wasm module.
+///
+/// * Parse that into your own `MyCustomSection` type in whatever way is
+///   appropriate.
+///
+/// * Do whatever kinds of inspection and manipulation of your `MyCustomSection`
+///   type you need to do.
+///
+/// * Use `my_module.customs.add(my_custom_section)` to add the custom section
+///   back into the module, so `walrus` can emit the processed/updated version
+///   of the custom section.
+#[derive(Debug, Default)]
+pub struct ModuleCustomSections {
+    arena: TombstoneArena<Option<Box<dyn CustomSectionAny>>>,
+}
+
+impl ModuleCustomSections {
+    /// Add a new custom section to the module.
+    pub fn add<T>(&mut self, custom_section: T) -> TypedCustomSectionId<T>
+    where
+        T: CustomSection,
+    {
+        let id = self
+            .arena
+            .alloc(Some(Box::new(custom_section) as Box<dyn CustomSectionAny>));
+        TypedCustomSectionId {
+            id: CustomSectionId(id),
+            _phantom: PhantomData,
+        }
+    }
+
+    /// Remove a custom section from the module.
+    pub fn delete<I>(&mut self, id: I)
```

Comment:
Perhaps `remove` to align with libstd collections?

Reply:
Our other APIs (eg for tables and functions, etc) use `delete`, so I think we should leave it in this PR, and maybe file a good-first-issue for switching all of them en mass.